TFLu: Allocate scratch tensors even though they are not in subgraph

mansnils opened the pull request:
This is fixing issue: https://github.com/tensorflow/tensorflow/issues/43537

mansnils committed:
TFLu: Allocate scratch tensors even though they are not in subgraph input

Change-Id: Iebe41ed254eeaef8e4ef6b2a2626291326c2a307

advaitjain commented:
@nkreeger, can you take a look?

nkreeger commented on tensorflow/lite/micro/micro_allocator_test.cc:
```diff
@@ -795,4 +795,71 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
 }
 
+TF_LITE_MICRO_TEST(TestTypicalFirstOpAndSecondOpWithScratchTensors) {
+  constexpr int nbr_tensors = 6;
```
nit: `number_tensors` - makes this a little easier to read when variables are not abbreviated.

mansnils replied:
Done

nkreeger commented on tensorflow/lite/micro/micro_allocator_test.cc:
```diff
@@ -749,14 +749,14 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
+  constexpr int t3 = 3;
+  constexpr int t4 = 4;
+
+  constexpr int num_conns = 2;
```
Same comment here - even though this is existing code, would be great if the variable name was spelled out to `number_connections` or similar.

mansnils replied:
Done

nkreeger commented on tensorflow/lite/micro/micro_allocator_test.cc:
```diff
@@ -795,4 +795,71 @@ TF_LITE_MICRO_TEST(TestOperatorInputsNotInSubgraphInputs) {
+  tflite::NodeAndRegistration* node_and_registration;
+  const int32_t metadata_buffer[tflite::testing::kOfflinePlannerHeaderSize +
+                                nbr_tensors] = {
+      1, 0, nbr_tensors,  // header: version, subgraph, nbr tensors
```
Google style guide generally suggests having inline comments like this:

```c++
  const int32_t metadata_buffer[tflite::testing::kOfflinePlannerHeaderSize +
                                nbr_tensors] = {/*version=*/1,
                                                /*subgraph=*/0,
                                                number_tensors,
                                                /*t0=*/0,
                                                /*t1=*/0,
                                                /*t2=*/0,
                                                /*t3=*/48,
                                                /*t4=*/-1};
```

mansnils replied:
Done

nkreeger reviewed (changes requested):
Looks mostly good - just some style cleanup.

mansnils committed:
TFlu: Fix bug in micro allocator test and update review comments

mansnils commented:
@nkreeger Ping for review

mansnils commented:
@advaitjain @nkreeger ping for review

nkreeger commented:
Hi @mansnils this looks like it doesn't build the micro_allocator_test:

```
tensorflow/lite/micro/micro_allocator_test.cc:836:70: error: too few arguments to function call, expected 3, have 2
      kTfLiteOk, allocator->FinishModelAllocation(model, eval_tensors));
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                    ^
./tensorflow/lite/micro/testing/micro_test.h:118:15: note: expanded from macro 'TF_LITE_MICRO_EXPECT_EQ'
    auto vy = y;                                                               \
              ^
./tensorflow/lite/micro/micro_allocator.h:137:16: note: 'FinishModelAllocation' declared here
  TfLiteStatus FinishModelAllocation(
               ^
1 error generated.
```

mansnils committed:
TFLu: Fix micro_allocator_test

FinishModelAllocation function interface was changed, on latest
upstream/master but not on the original topic branch.
